core/src/main/java/org/elasticsearch/common/geo/GeoDistance.java

```diff
@@ -127,12 +130,27 @@ public FixedSourceDistance fixedSourceDistance(double sourceLatitude, double sou
         }
     };
 
+    /** Returns a GeoDistance object as read from the StreamInput. */
+    @Override
+    public GeoDistance readFrom(StreamInput in) throws IOException {
+        return GeoDistance.values()[in.readVInt()];
```

Comment:
We should throw an IOException with a nice message if the read int is greater than `values().length`. See ShapeRelation for example of what I mean. We also need unit tests to make sure the ordinals don't change. See ShapeRelationTests for an example of that.